apollo-runtime/src/main/java/com/apollographql/apollo/internal/RealApolloCall.java

```diff
@@ -210,61 +211,69 @@ private RealApolloCall(Builder<T> builder) {
         .build();
   }
 
-  final class AsyncCall implements ApolloInterceptor.CallBack {
-
-    private final Callback<T> responseCallback;
-
-    private AsyncCall(Callback<T> callback) {
-      this.responseCallback = callback;
-    }
-
-    @Override public void onResponse(@Nonnull final ApolloInterceptor.InterceptorResponse response) {
-      try {
+  @Nonnull @Override public Operation operation() {
+    return operation;
+  }
 
-        if (responseCallback == null) {
-          return;
-        }
-        if (canceled) {
-          responseCallback.onCanceledError(new ApolloCanceledException("Canceled"));
-          return;
-        }
-        if (queryFetcher.isPresent()) {
-          queryFetcher.get().refetchAsync(new QueryFetcher.OnFetchCompleteCallback() {
-            @Override public void onFetchComplete() {
-              //noinspection unchecked
-              responseCallback.onResponse(response.parsedResponse.get());
-            }
-          });
-        } else { //noinspection unchecked
-          responseCallback.onResponse(response.parsedResponse.get());
+  private void refetchQueryWatchers() {
+    try {
+      for (OperationName operationName : refetchQueryNames) {
+        for (ApolloQueryWatcher queryWatcher : tracker.activeQueryWatchers(operationName)) {
+          queryWatcher.refetch();
         }
-      } finally {
-        tracker.onAsyncCallFinished(this);
       }
+    } catch (Exception e) {
+      logger.e(e, "Failed to re-fetch query watcher");
     }
+  }
 
-    @Override public void onFailure(@Nonnull ApolloException e) {
-      try {
-
-        if (responseCallback == null) {
-          return;
+  private ApolloInterceptor.CallBack interceptorCallbackProxy(final Callback<T> originalCallback) {
+    return new ApolloInterceptor.CallBack() {
+      @Override public void onResponse(@Nonnull final ApolloInterceptor.InterceptorResponse response) {
+        if (originalCallback == null) return;
+        try {
+          if (RealApolloCall.this.canceled) {
+            originalCallback.onCanceledError(new ApolloCanceledException("Canceled"));
+            return;
+          }
+
+          refetchQueryWatchers();
+
+          if (queryFetcher.isPresent()) {
+            queryFetcher.get().refetchAsync(new QueryFetcher.OnFetchCompleteCallback() {
+              @Override public void onFetchComplete() {
+                //noinspection unchecked
+                originalCallback.onResponse(response.parsedResponse.get());
```

Comment:
Would it be safe to not block the original call on the refetch queries? I can see the argument for both ways, ideally if it blocked or not would be user configurable.

Additionally, the actually refetch routine seems to be serially executed. Maybe we should investigate just dumping all the refetch querries into one graphql query? Although is may complicate query whitelisting.